### PR TITLE
[6.5] [Metadata] Getting complex-type metadata

### PR DESCRIPTION
**Pull request by pkosko**
[6.5] [Metadata] Getting complex-type metadata - adding API reference
[6.5] [Metadata] Getting complex-type metadata - adding guide

ACR: https://code.sec.samsung.net/jira/browse/TWDAPI-271

### CODE REVIEW

File: docs/application/web/guides/multimedia/metadata.md [hunk 1]
```diff
@@ -2,8 +2,27 @@
 
 You can extract metadata information from multimedia files. Metadata is the information about multimedia files.
 
-The main feature of the Metadata API is to extract metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.
+The main feature of the Metadata API include:
```
Comment (rwalczyna) [inline docs/application/web/guides/multimedia/metadata.md:5]:
```suggestion
The main features of the Metadata API include:
```

File: docs/application/web/guides/multimedia/metadata.md [hunk 2]
```diff
@@ -44,6 +63,105 @@ When all the needed metadata is extracted, the file handle needs to be released
...
+   var fileHandle = tizen.metadata.createFileHandle(filePath);
+   ```
+
+2. When the file handle is properly created, artwork can be extracted:
+
+   ```javascript
+   var artwork = fileHandle.getArtwork();
+   ```
+
+3. *Blob* object can be easily shown html code in your application.
```
Comment (rwalczyna) [inline docs/application/web/guides/multimedia/metadata.md:81]: Can you rephrase this sentence? It doesn't sound clear. (It occurs 3 times in this file)

**Review by rwalczyna — CHANGES_REQUESTED** [review]

File: docs/application/web/guides/multimedia/metadata.md (hunk 2 above)
Comment (pkosko) [reply docs/application/web/guides/multimedia/metadata.md:81]:
```suggestion
3. *Blob* object can be easily shown in your application in html *image* tag.
```
will this be ok?

Comment (rwalczyna) [reply docs/application/web/guides/multimedia/metadata.md:81]: Yes, but remember that '<' has to be escaped in md - by '\<' or '&lt;'.

**Review by rwalczyna — APPROVED** [review]
LGTM

File: docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html [hunk 3]
```diff
@@ -80,9 +82,21 @@ <h2 id="method-summary">Summary of Interfaces and Methods</h2>
 <td><a href="#MetadataFileHandle">MetadataFileHandle</a></td>
 <td>
 <div>DOMString <a href="#MetadataFileHandle::get">get</a> (<a href="#MetadataType">MetadataType</a> type)</div>
+<div>
+<a href="filesystem.html#Blob">Blob</a> <a href="#MetadataFileHandle::getArtwork">getArtwork</a> ()</div>
+<div>
+<a href="filesystem.html#Blob">Blob</a> <a href="#MetadataFileHandle::getThumbnailFrame">getThumbnailFrame</a> ()</div>
```
Comment (mijin85cho) [inline docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html:88]: Just a question regarding https://code.sec.samsung.net/jira/browse/TWDAPI-271.

In the guide description, below are added
 - getArtwork()
 - getThumbnailFrame()
 - getFrameAtTime()
 - getSyncLyrics()

However from the list in the JIRA, I couldn't find getThumbnailFrame. Is there any reason for that?

**Review by mijin85cho — COMMENTED** [review]
@annie-samsung, when you start reviewing or assign the reviewer, please refer below points also.

File: docs/application/web/api/6.5/device_api/tv/tizen/metadata.html [hunk 4]
```diff
@@ -52,6 +52,8 @@ <h2>Table of Contents</h2>
 </li>
```
[the remark this answers is not in the document]
Comment (pkosko) [reply docs/application/web/api/6.5/device_api/tv/tizen/metadata.html:52]: Privacy related privileges (and PPM module) are not supported on TV devices, only mobile and wearable.

[the remark this answers is not in the document]
Comment (pkosko) [reply docs/application/web/api/6.5/device_api/tv/tizen/metadata.html:52]: Privacy related privileges (and PPM Web API) are not supported on TV, thus this comment is removed as not related.

File: docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html (hunk 3 above)
Comment (pkosko) [reply docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html:88]: You are right, the name was changed during review, but I didn't update ACR description. I fixed description of https://code.sec.samsung.net/jira/browse/TWDAPI-271

File: docs/application/web/guides/multimedia/metadata.md [hunk 5]
```diff
@@ -44,6 +63,105 @@ When all the needed metadata is extracted, the file handle needs to be released
...
+   var fileHandle = tizen.metadata.createFileHandle(filePath);
+   ```
+
+2. When the file handle is properly created, one need to check how much lyrics are included.
+
+   ```javascript
+   var lyricsNum = fileHandle.get("SYNCLYRICS_NUM");
+   ```
+
+3. Lyrics then can be gathered and printed:
```
Comment (annie-samsung) [inline docs/application/web/guides/multimedia/metadata.md:155]: Gather and print the lyrics.

File: docs/application/web/guides/multimedia/metadata.md [hunk 6]
```diff
@@ -44,6 +63,105 @@ When all the needed metadata is extracted, the file handle needs to be released
...
+
+   ```javascript
+   for (var i = 0; i < lyricsNum; ++i) {
+      var lyrics = fileHandle.getSyncLyrics(i);
+      console.log("Lyrics " + i + " (" + lyrics.timestamp  + "ms): " + lyrics.lyrics);
+   }
+   ```
+
+4. After use, remember to release fileHandle.
```
Comment (annie-samsung) [inline docs/application/web/guides/multimedia/metadata.md:164]: 4. After use, release fileHandle.

File: docs/application/web/guides/multimedia/metadata.md [hunk 7]
```diff
@@ -44,6 +63,105 @@ When all the needed metadata is extracted, the file handle needs to be released
...
+   ```
+
+3. *Blob* object can be easily shown in your application in html *&lt;img&gt;* tag
+
+   ```javascript
+   var elem = document.getElementById("blobImage");
+   elem.src = URL.createObjectURL(frame);
+   ```
+
+4. After use, remember to release fileHandle.
```
Comment (annie-samsung) [inline docs/application/web/guides/multimedia/metadata.md:138]: 4. After use, release fileHandle.

File: docs/application/web/guides/multimedia/metadata.md [hunk 8]
```diff
@@ -44,6 +63,105 @@ When all the needed metadata is extracted, the file handle needs to be released
...
+
+## Extract lyrics
+
+1. Create a file handle using the `createFileHandle()` method. Lyrics need to be included in multimedia file as SYLT metadata tag. Supported formats of encoding are at least <a href="http://www.ietf.org/rfc/rfc2279.txt">UTF-8</a> and <a href="http://en.wikipedia.org/wiki/ISO/IEC_8859-1">ISO-8859-1</a>. Other formats can be ignored depending on the platform.
```
Comment (annie-samsung) [inline docs/application/web/guides/multimedia/metadata.md:142]: This part can be added to introductory sentence with a little bit of repharsing:
Lyrics need to be included in multimedia file a...

File: docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html [hunk 9]
```diff
@@ -382,6 +400,209 @@ <h4>Methods</h4>
...
+                </p></li>
+<li class="list"><p>
+ with error type AbortError, if any other error occurs.
+                </p></li>
+</ul>
+</li></ul>
+        </div>
+<div class="example">
+<span class="example"><p>Code example:</p></span><pre name="code" class="examplecode prettyprint">/* precondition - path needs to point to a file with a non-empty artwork */
+var path = "music/sample.mp3";
```
Comment (annie-samsung) [inline docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html:435]: music/sample.mp3 is that correct?

**Review by annie-samsung — COMMENTED** [review]
@pkosko,
please find my comments. Based on your responses and fixes I may have additional comments.
Where ever the content is repetitively appearing, please treat those comments as global and fix them everywhere (if applicable).

**Comment by pkosko** [comment]
I've uploaded changes in API reference https://github.com/Samsung/tizen-docs/pull/1408/commits/ee8f2e7be17fd44cadfcecbc9a110a3478b3c1d9 , fixing guide is in progress yet.

File: docs/application/web/guides/multimedia/metadata.md [hunk 10]
```diff
@@ -2,8 +2,27 @@
...
-The main feature of the Metadata API is to extract metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.
+The main features of the Metadata API include:
 
+- Extracting simple metadata informations
+
+  You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.
+
+- Extracting artwork of a multimedia file
+
+  You can [extract artwork](#extract-artwork) from multimedia files using the `getArtwork()` method to get *Blob* object containing the artwork.
```
[the remark this answers is not in the document]
Comment (mijin85cho) [reply docs/application/web/guides/multimedia/metadata.md:13]: In tizen-docs repository, we don't recommend italic font. How about changing it to bold?

File: docs/application/web/guides/multimedia/metadata.md [hunk 11]
```diff
@@ -44,6 +63,105 @@ When all the needed metadata is extracted, the file handle needs to be released
...
+## Extract certain frame
+
+1. Create a file handle using the `createFileHandle()` method for a video file.
+
+   ```javascript
+   var filePath = "videos/sample_video.mp4";
+   var fileHandle = tizen.metadata.createFileHandle(filePath);
+   ```
+
+2. When the file handle is properly created, thumbnail frame can be extracted. You need to provide timestamp information (in milliseconds) and flag indicating to return frame from exact given time (true) or the user can only get the nearest <a href="https://en.wikipedia.org/wiki/Video_compression_picture_types">I-frame</a> of the video (false). Gathering nearest I-frame has better performance.
```
[the remark this answers is not in the document]
Comment (pkosko) [reply docs/application/web/guides/multimedia/metadata.md]: This sentence is strictly related to parameters which are passed to getFramteAtTime function - according to this, I think that this is correct place for this information.
I rephrased this sentence the same way as in API reference.

File: docs/application/web/guides/multimedia/metadata.md (hunk 8 above)
Comment (pkosko) [reply docs/application/web/guides/multimedia/metadata.md]: This is a first point of a scenario, I think that it is in correct place, as it describes conditions regarding validity of a file for this scenario.

**Comment by pkosko** [comment]
@annie-samsung I think that I applied/responded to all of your comments.

File: docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html (hunk 9 above)
Comment (annie-samsung) [reply docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html:435]: okay

File: docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html [hunk 12]
```diff
@@ -382,6 +400,209 @@ <h4>Methods</h4>
...
+<div class="parameters">
+<p><span class="param">Parameters:</span></p>
+<ul>
+          <li class="param">
+<span class="name">timestamp</span>:
+ A timestamp of a frame to gather in milliseconds.
+                </li>
+          <li class="param">
+<span class="name">isAccurate</span><span class="optional"> [optional]</span><span class="optional"> [nullable]</span>:
+ A flag indicating if need to be frame from exact given time (true) or the user can only get the nearest <a href="https://en.wikipedia.org/wiki/Video_compression_picture_types">I-frame</a> of the video (false). Gathering nearest I-frame has better performance.
```
[the remark this answers is not in the document]
Comment (annie-samsung) [reply docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html]: Thanks looks good.

File: docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html [hunk 13]
```diff
@@ -423,6 +644,45 @@ <h4>Methods</h4>
 </dl>
 </div>
 </div>
+<div class="interface" id="MetadataSyncLyrics">
+<a class="backward-compatibility-anchor" name="::Metadata::MetadataSyncLyrics"></a><h3>2.4. MetadataSyncLyrics</h3>
+<div class="brief">
+ Object representing synchronized lyrics for a certain point of time.
```
[the remark this answers is not in the document]
Comment (annie-samsung) [reply docs/application/web/api/6.5/device_api/mobile/tizen/metadata.html]: Thanks for clarifying and fixing it.

File: docs/application/web/guides/multimedia/metadata.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (annie-samsung) [reply docs/application/web/guides/multimedia/metadata.md]: Thanks for pointing it out.

**Review by annie-samsung — COMMENTED** [review]
Thanks for fixing the first set of comments, and also for clarifying my queries.
I suppose these are the final set of comments that I have.

**Comment by pkosko** [comment]
@annie-samsung I've applied all your comments. In case of merging this PR consider using 'Squash and merge' merging strategy to avoid flooding main branch with all review-fixes-commits and just do it as single commit.

File: docs/application/web/guides/multimedia/metadata.md [hunk 14]
```diff
@@ -23,27 +42,126 @@ Additionally, to access files using the Metadata API (in [mobile](../../api/late
...
+## Extract certain frame
+
+1. Create a file handle using `createFileHandle()` for a video file.
+
+   ```javascript
+   var filePath = "videos/sample_video.mp4";
+   var fileHandle = tizen.metadata.createFileHandle(filePath);
+   ```
+
+2. When the file handle is created, thumbnail frame can be extracted. You need to provide timestamp information (in milliseconds) and a flag indicating accuracy. When the flag value is set to 'true', flag indicates extracting an exact frame for the given time. When the value is 'false', flag indicates extracting an <a href="https://en.wikipedia.org/wiki/Video_compression_picture_types">I-frame</a> nearest to the given time. Gathering nearest I-frame has better performance.
```
Comment (annie-samsung) [inline docs/application/web/guides/multimedia/metadata.md:123]: 'true' 
'false'
Was the intention to use code tags? if yes please change to:


`true`  
`false`

File: docs/application/web/guides/multimedia/metadata.md [hunk 15]
```diff
@@ -23,27 +42,126 @@ Additionally, to access files using the Metadata API (in [mobile](../../api/late
...
+## Extract lyrics
+
+1. Create a file handle using `createFileHandle()`. Lyrics need to be included in multimedia file as SYLT metadata tag. Supported formats of encoding are at least <a href="http://www.ietf.org/rfc/rfc2279.txt">UTF-8</a> and <a href="http://en.wikipedia.org/wiki/ISO/IEC_8859-1">ISO-8859-1</a>. Other formats can be ignored depending on the platform.
+
+   ```javascript
+   var filePath = "music/sample.mp3";
+   var fileHandle = tizen.metadata.createFileHandle(filePath);
+   ```
+
+2. When the file handle is created, one need to check how many lyrics are included.
```
Comment (annie-samsung) [inline docs/application/web/guides/multimedia/metadata.md:149]: In my previous comment, I had removed "one need to" as it is implicit that the developer needs to check this. Secondly, in technical documentation, we use second-person pronouns such you, your, etc., and avoid using third-person pronouns completely. 

Is there any reason you retained it?
If it is was an oversight could you change to:

2. When the file handle is created, check how many lyrics are included.

**Review by annie-samsung — COMMENTED** [review]
Hello @pkosko, these should be my final set of comments.

**Comment by annie-samsung** [comment]
> @annie-samsung I've applied all your comments. In case of merging this PR consider using 'Squash and merge' merging strategy to avoid flooding main branch with all review-fixes-commits and just do it as single commit.

Sure I will do Squash and merge at the time of merging this PR. Thanks for the suggestion.

File: docs/application/web/guides/multimedia/metadata.md (hunk 14 above)
Comment (pkosko) [reply docs/application/web/guides/multimedia/metadata.md]: done

File: docs/application/web/guides/multimedia/metadata.md (hunk 15 above)
Comment (pkosko) [reply docs/application/web/guides/multimedia/metadata.md]: Yes, I've missed this part. Sometimes it is easy to miss some changes in sentence (especially, if there are many comments). Maybe you should consider using github built-in 'suggestion' mechanism in your comments - it is much easier then to apply a suggestion with just one click without any manual copy-paste.

**Review by annie-samsung — APPROVED** [review]
Thanks for fixing all the comments, looks good hence approving.
